wpiutil/src/main/native/include/wpi/uv/AsyncFunction.h

```diff
@@ -0,0 +1,238 @@
+        // Yes, remove it from the vector and we're done.
+        m_results.erase(it);
+        return;
+      }
+      // No, keep waiting for a response
+      m_resultCv.wait(lock);
+    }
+  }
+
+ private:
+  std::vector<std::thread::id> m_results;
+};
+
+}  // namespace detail
+
+template <typename T>
+class AsyncFunction;
+
+/**
+ * Function async handle.
+ * Async handles allow the user to "wakeup" the event loop and have a function
+ * called from another thread that returns a result to the calling thread.
+ */
+template <typename R, typename... T>
+class AsyncFunction<R(T...)> final
+    : public HandleImpl<AsyncFunction<R(T...)>, uv_async_t>,
+      private detail::AsyncFunctionHelper<R, T...> {
+  struct private_init {};
+
+ public:
+  AsyncFunction(std::function<R(T...)> func, const private_init&)
+      : wakeup{func} {}
+  ~AsyncFunction() noexcept override = default;
+
+  /**
+   * Create an async handle.
+   *
+   * @param loop Loop object where this handle runs.
+   * @param func wakeup function to be called (sets wakeup value)
+   */
+  static std::shared_ptr<AsyncFunction> Create(
+      Loop& loop, std::function<R(T...)> func = nullptr) {
+    auto h = std::make_shared<AsyncFunction>(std::move(func), private_init{});
+    int err = uv_async_init(loop.GetRaw(), h->GetRaw(), [](uv_async_t* handle) {
+      AsyncFunction& h = *static_cast<AsyncFunction*>(handle->data);
+      std::lock_guard<wpi::mutex> lock(h.m_mutex);
+
+      if (!h.m_params.empty()) {
+        // for each set of parameters in the input queue
+        for (auto&& v : h.m_params) {
+          // call the wakeup function and put the result in the output queue
+          // if the caller is waiting for it
+          if (v.first == std::thread::id{})
+            apply_tuple(h.wakeup, std::move(v.second));
+          else
+            h.RunCall(h.wakeup, v);
+        }
+        h.m_params.clear();
+        // wake up any threads that might be waiting for the result
+        h.m_resultCv.notify_all();
+      }
+    });
+    if (err < 0) {
+      loop.ReportError(err);
+      return nullptr;
+    }
+    h->Keep();
+    return h;
+  }
+
+  /**
+   * Create an async handle.
+   *
+   * @param loop Loop object where this handle runs.
+   * @param func wakeup function to be called (sets wakeup value)
+   */
+  static std::shared_ptr<AsyncFunction> Create(
+      const std::shared_ptr<Loop>& loop,
+      std::function<R(T...)> func = nullptr) {
+    return Create(*loop, std::move(func));
+  }
+
+  /**
+   * Wakeup the event loop, call the async function, and ignore any result.
+   * This is non-blocking and does NOT wait until the async function returns.
+   *
+   * It’s safe to call this function from any thread EXCEPT the loop thread.
+   * The async function will be called on the loop thread.
+   */
+  template <typename... U>
+  void Send(U&&... u) {
+    // add the parameters to the input queue
+    {
+      std::lock_guard<wpi::mutex> lock(this->m_mutex);
```

Comment:
Isn't `this->` redundant here? (Same with a few other instances.)

Reply:
Nope, it's required because of the way templates work (e.g. you get a compiler error without it).